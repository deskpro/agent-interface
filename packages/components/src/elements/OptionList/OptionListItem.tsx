import * as React from "react";
import classNames from "classnames";
import { Manager, Reference, Popper } from "react-popper";

import { MenuProps } from "../Menu/Menu";
import Checkbox from "../../inputs/Checkbox/Checkbox";
import Icon from "../Icon/Icon";
import useMenu from "../../hooks/useMenu";

export type OptionListItemProps = {
  name: React.Key;
  title: string;
  disabled?: boolean;
  checkable?: boolean;
  checked?: boolean;
  onCheck?: (itemName: React.Key, checked: boolean) => void;
  renderGearMenu?: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
  isDragging?: boolean;
  draggableProps?: any;
  dragHandleProps?: any;
  innerRef?: React.Ref<HTMLElement>;
};

const OptionListItem: React.FC<OptionListItemProps> = ({
  name,
  title,
  disabled,
  checkable,
  checked,
  onCheck,
  renderGearMenu,
  isDragging = false,
  draggableProps = {},
  dragHandleProps = {},
  innerRef
}) => {
  const { isVisible: menuIsVisible, toggleMenu } = useMenu();

  return (
    <div
      className={classNames("dp-OptionList-item dp-Menu-linkItem", {
        "dp-Uneditable": disabled,
        "is-dragging": isDragging
      })}
      ref={innerRef}
      {...draggableProps}
    >
      {!!dragHandleProps && <Icon name="drag" size={12} {...dragHandleProps} />}
      {checkable && (
        <Checkbox
          disabled={disabled}
          id={name as string}
          name={name as string}
          checked={checked}
          onChange={(e: React.FormEvent) => {
            if (onCheck) {
              const {
                name: itemName,
                checked: isItemChecked
              } = e.target as HTMLInputElement;
              onCheck(itemName, isItemChecked);
            }
          }}
        />
      )}
      {title}

      {!!renderGearMenu && (
        <Manager>
          <Reference>
            {({ ref }) => (
              <Icon name="settings" size={12} onClick={toggleMenu} ref={ref} />
            )}
          </Reference>
          {menuIsVisible && (
            <Popper
              placement="right-start"
              modifiers={{
                preventOverflow: {
                  enabled: true,
                  escapeWithReference: true,
                  boundariesElement: "viewport"
                },
                flip: {
                  enabled: true,
                  flipVariationsByContent: true
                }
              }}
            >
              {({ ref, style }) => renderGearMenu({ menuRef: ref, style })}
            </Popper>
          )}
        </Manager>
      )}
    </div>
  );
};

export default OptionListItem;
