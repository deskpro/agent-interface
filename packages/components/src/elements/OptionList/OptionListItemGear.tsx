import * as React from "react";
import { Manager, Reference, Popper } from "react-popper";

import Icon from "../Icon/Icon";
import { MenuProps } from "../Menu/Menu";
import useMenu from "../../hooks/useMenu";
import useOutsideClick from "../../hooks/useOutsideClick";

export type OptionListItemGearProps = {
  renderGearMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
};

const OptionListItemGear: React.FC<OptionListItemGearProps> = ({
  renderGearMenu
}) => {
  const { isVisible: menuIsVisible, toggleMenu, hideMenu } = useMenu();

  const gearMenuRef = React.useRef<HTMLSpanElement>(null);
  useOutsideClick(gearMenuRef, hideMenu);

  return (
    <span className="dp-OptionList-itemGearMenu" ref={gearMenuRef}>
      <Manager>
        <Reference>
          {({ ref }) => (
            <Icon name="settings" size={12} onClick={toggleMenu} ref={ref} />
          )}
        </Reference>
        {menuIsVisible && (
          <Popper
            placement="right-start"
            modifiers={{
              offset: { offset: "-1,5" },
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              },
              flip: {
                enabled: true,
                flipVariationsByContent: true
              }
            }}
          >
            {({ ref, style }) =>
              renderGearMenu({ menuRef: ref, style, onMenuClose: hideMenu })
            }
          </Popper>
        )}
      </Manager>
    </span>
  );
};

export default OptionListItemGear;
