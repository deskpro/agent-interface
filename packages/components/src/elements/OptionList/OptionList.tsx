import * as React from "react";
import classNames from "classnames";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";

import OptionListItem from "./OptionListItem";

type OptionListSubcomponents = {
  Item: typeof OptionListItem;
};

export type OptionListProps = {
  className?: string;
  style?: any;
  name: React.Key;
  sortable?: boolean;
  onOrderChange?: (
    itemName: React.Key,
    oldPosition: number,
    newPosition: number
  ) => void;
  checkable?: boolean;
  children: React.ReactElement<any>[];
};

const OptionList: React.FC<OptionListProps> & OptionListSubcomponents = ({
  className,
  name,
  style,
  children,
  checkable,
  sortable,
  onOrderChange
}) => {
  const onDragEnd = React.useCallback(
    ({ draggableId, source, destination }) => {
      if (onOrderChange) {
        onOrderChange(draggableId, source.index, destination.index);
      }
    },
    [onOrderChange]
  );

  return (
    <div
      className={classNames("dp-OptionList dp-Menu", className)}
      style={style}
    >
      {!sortable &&
        React.Children.map(children, (child: React.ReactElement<any>) =>
          React.cloneElement(child, { checkable })
        )}
      {sortable && (
        <DragDropContext onDragEnd={onDragEnd}>
          <Droppable droppableId={name} type={name}>
            {(provided, snapshot) => (
              <span ref={provided.innerRef}>
                {React.Children.map(
                  children,
                  (child: React.ReactElement<any>, index) => (
                    <Draggable
                      draggableId={child.props.name}
                      index={index}
                      isDragDisabled={child.props.disabled}
                    >
                      {draggableProvided =>
                        React.cloneElement(child, {
                          checkable,
                          showDragHandle: true,
                          isDragging: snapshot.isDragging,
                          ...draggableProvided
                        })
                      }
                    </Draggable>
                  )
                )}
              </span>
            )}
          </Droppable>
        </DragDropContext>
      )}
    </div>
  );
};

OptionList.Item = OptionListItem;

export default OptionList;
