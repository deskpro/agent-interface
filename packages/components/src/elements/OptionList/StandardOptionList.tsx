import * as React from "react";
import OptionList from "./OptionList";
import Menu from "../Menu/Menu";

export type OptionItemSettings = {
  name: React.Key;
  text: string;
};

export type OptionItemType = {
  name: React.Key;
  title: string;
  disabled?: boolean;
  checked?: boolean;
  settings?: OptionItemSettings[];
};

export type StandardOptionListProps = {
  name: string;
  className?: string;
  items: OptionItemType[];
  sortable?: boolean;
  onOrderChange?: (
    itemName: React.Key,
    oldPosition: number,
    newPosition: number
  ) => void;
  checkable?: boolean;
  onCheck?: (itemName: React.Key, checked: boolean) => void;
  onSettingChange?: (settingName: React.Key, checked: boolean) => void;
};

const StandardOptionList: React.FC<StandardOptionListProps> = ({
  items,
  onCheck,
  onSettingChange,
  ...props
}) => (
  <OptionList {...props}>
    {items.map(({ settings, ...item }) => (
      <OptionList.Item
        {...item}
        key={item.name}
        onCheck={onCheck}
        renderGearMenu={
          settings
            ? menuProps => (
                <Menu
                  {...menuProps}
                  isCheckable
                  onItemCheck={onSettingChange}
                  checkedItems={[settings[0].name]}
                >
                  {settings.map(setting => (
                    <Menu.MenuItem {...setting} key={setting.name} />
                  ))}
                </Menu>
              )
            : undefined
        }
      />
    ))}
  </OptionList>
);

export default StandardOptionList;
