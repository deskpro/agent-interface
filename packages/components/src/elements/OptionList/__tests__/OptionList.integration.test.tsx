import * as React from "react";
import { mount } from "enzyme";

import OptionList from "../OptionList";

describe("<OptionList />", () => {
  it("should match its snapshot", () => {
    const wrapper = mount(
      <OptionList name="fields" checkable>
        <OptionList.Item name="agent" title="Agent" disabled checked />
        <OptionList.Item name="title" title="Title" disabled checked />
        <OptionList.Item name="id" title="ID" />
      </OptionList>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
