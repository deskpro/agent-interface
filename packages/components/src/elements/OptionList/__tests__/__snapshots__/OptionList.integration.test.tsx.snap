// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptionList /> should match its snapshot 1`] = `
<OptionList
  checkable={true}
  name="fields"
>
  <div
    className="dp-OptionList dp-Menu"
  >
    <OptionListItem
      checkable={true}
      checked={true}
      disabled={true}
      key=".0"
      name="agent"
      title="Agent"
    >
      <div
        className="dp-OptionList-item dp-Menu-linkItem dp-Uneditable"
      >
        <Checkbox
          checked={true}
          disabled={true}
          id="agent"
          label=""
          name="agent"
          onChange={[Function]}
        >
          <span
            className="dp-CustomCheckbox"
          >
            <input
              checked={true}
              className="dp-ControlInput"
              disabled={true}
              id="agent"
              name="agent"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className=""
              htmlFor="agent"
            />
          </span>
        </Checkbox>
        Agent
      </div>
    </OptionListItem>
    <OptionListItem
      checkable={true}
      checked={true}
      disabled={true}
      key=".1"
      name="title"
      title="Title"
    >
      <div
        className="dp-OptionList-item dp-Menu-linkItem dp-Uneditable"
      >
        <Checkbox
          checked={true}
          disabled={true}
          id="title"
          label=""
          name="title"
          onChange={[Function]}
        >
          <span
            className="dp-CustomCheckbox"
          >
            <input
              checked={true}
              className="dp-ControlInput"
              disabled={true}
              id="title"
              name="title"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className=""
              htmlFor="title"
            />
          </span>
        </Checkbox>
        Title
      </div>
    </OptionListItem>
    <OptionListItem
      checkable={true}
      key=".2"
      name="id"
      title="ID"
    >
      <div
        className="dp-OptionList-item dp-Menu-linkItem"
      >
        <Checkbox
          id="id"
          label=""
          name="id"
          onChange={[Function]}
        >
          <span
            className="dp-CustomCheckbox"
          >
            <input
              className="dp-ControlInput"
              id="id"
              name="id"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className=""
              htmlFor="id"
            />
          </span>
        </Checkbox>
        ID
      </div>
    </OptionListItem>
  </div>
</OptionList>
`;
